File: src/client-gateway/android/TollDNSGateway.kt
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

class TollDNSGateway(private val resolverUrl: String) {
    fun resolve(queryName: String, queryType: String, voucherJson: String): String {
        val url = URL("$resolverUrl/v1/resolve")
        val connection = url.openConnection() as HttpURLConnection
        connection.requestMethod = "POST"
        connection.setRequestProperty("Content-Type", "application/json")
        connection.doOutput = true

        val payload = """{
            \"voucher\": $voucherJson,
            \"query\": {\"name\": \"$queryName\", \"type\": \"$queryType\", \"needsGateway\": true}
        }"""

        OutputStreamWriter(connection.outputStream).use { it.write(payload) }
        return connection.inputStream.bufferedReader().use { it.readText() }
    }
}

File: src/client-gateway/ios/TollDNSGateway.swift
import Foundation

struct TollDNSGateway {
    let resolverURL: URL

    func resolve(queryName: String, queryType: String, voucher: [String: Any]) async throws -> [String: Any] {
        var request = URLRequest(url: resolverURL.appendingPathComponent("/v1/resolve"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let payload: [String: Any] = [
            "voucher": voucher,
            "query": [
                "name": queryName,
                "type": queryType,
                "needsGateway": true
            ]
        ]

        request.httpBody = try JSONSerialization.data(withJSONObject: payload)
        let (data, _) = try await URLSession.shared.data(for: request)
        let response = try JSONSerialization.jsonObject(with: data)
        return response as? [String: Any] ?? [:]
    }
}
